into/backends/aws.py

```diff
@@ -0,0 +1,248 @@
+
+    if aws_access_key_id is None:
+        aws_access_key_id = os.environ.get('AWS_ACCESS_KEY_ID')
+
+    if aws_secret_access_key is None:
+        aws_secret_access_key = cfg.get('Credentials', 'aws_secret_access_key')
+
+    if aws_secret_access_key is None:
+        aws_secret_access_key = os.environ.get('AWS_SECRET_ACCESS_KEY')
+
+    # anon is False but we didn't provide any credentials so try anonymously
+    anon = (not anon and
+            aws_access_key_id is None and
+            aws_secret_access_key is None)
+    return boto.connect_s3(aws_access_key_id, aws_secret_access_key,
+                           anon=anon)
+
+
+class _S3(object):
+    def __init__(self, uri, s3=None, aws_access_key_id=None,
+                 aws_secret_access_key=None, *args, **kwargs):
+        result = urlparse(uri)
+        self.bucket = result.netloc
+        self.key = result.path.lstrip('/')
+
+        if s3 is not None:
+            self.s3 = s3
+        else:
+            self.s3 = get_s3_connection(aws_access_key_id=aws_access_key_id,
+                                        aws_secret_access_key=aws_secret_access_key)
+        try:
+            bucket = self.s3.get_bucket(self.bucket)
+        except boto.exception.S3ResponseError:
+            bucket = self.s3.create_bucket(self.bucket)
+
+        self.object = bucket.get_key(self.key)
+        if self.object is None:
+            self.object = bucket.new_key(self.key)
+
+        self.subtype.__init__(self, uri, *args, **kwargs)
+
+
+@memoize
+def S3(cls):
+    """Parametrized S3 bucket Class
+
+    Examples
+    --------
+    >>> import sqlalchemy
+    >>> S3(CSV)
+    S3(CSV)
+    >>> S3(sqlalchemy.Table)
+    """
+    return type('S3(%s)' % cls.__name__, (_S3, cls), {'subtype': cls})
+
+
+@sample.register((S3(CSV), S3(JSONLines)))
+@contextmanager
+def sample_s3_line_delimited(data, length=8192):
+    """Get bytes from `start` to `stop` of `key` from the S3 bucket `bucket`.
+
+    Parameters
+    ----------
+    data : S3(CSV)
+        A CSV file living in  an S3 bucket
+    length : int, optional, default ``8192``
+        Number of bytes of the file to read
+
+    """
+    headers = {'Range': 'bytes=0-%d' % length}
+    raw = data.object.get_contents_as_string(headers=headers)
+
+    # this is generally cheap as we usually have a tiny amount of data
+    try:
+        index = raw.rindex(b'\r\n')
+    except ValueError:
+        index = raw.rindex(b'\n')
+
+    raw = raw[:index]
+
+    with tmpfile(ext(data.path)) as fn:
+        # we use wb because without an encoding boto returns bytes
+        with open(fn, 'wb') as f:
+            f.write(raw)
+        yield fn
+
+
+@discover.register((S3(CSV), S3(JSONLines)))
+def discover_s3_line_delimited(c, length=8192, **kwargs):
+    """Discover CSV and JSONLines files from S3.
+
+    Examples
+    --------
+    >>> csv_dshape = discover(resource('s3://nyqpug/tips.csv'))
+    >>> print(csv_dshape)
```

Comment:
Nope, this is a public read-only S3 bucket that I manage. This should only fail if I get hit by a :bus: